drivers/eeprom/eeprom_mchp_xec.c

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2022 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_xec_eeprom
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/eeprom.h>
+#include <soc.h>
+
+#ifdef CONFIG_PINCTRL
+#include <zephyr/drivers/pinctrl.h>
+#endif
+
+#define LOG_LEVEL CONFIG_EEPROM_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(eeprom_xec);
+
+/* EEPROM Mode Register */
+#define XEC_EEPROM_MODE_ACTIVATE			BIT(0)
+
+/* EEPROM Status Register */
+#define XEC_EEPROM_STS_TRANSFER_COMPL			BIT(0)
+
+/* EEPROM Execute Register - Transfer size bit position */
+#define XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS		(24)
+
+/* EEPROM Execute Register - Commands */
+#define XEC_EEPROM_EXC_CMD_READ			0x00000
+#define XEC_EEPROM_EXC_CMD_WRITE		0x10000
+#define XEC_EEPROM_EXC_CMD_READ_STS		0x20000
+#define XEC_EEPROM_EXC_CMD_WRITE_STS	0x30000
+
+/* EEPROM Execute Register - Address mask */
+#define XEC_EEPROM_EXC_ADDR_MASK		0x7FFU
+
+/* EEPROM Read/Write Transfer Size */
+#define XEC_EEPROM_TRANSFER_SIZE_READ	32
+#define XEC_EEPROM_TRANSFER_SIZE_WRITE	32
+
+#define XEC_EEPROM_PAGE_SIZE			32
+
+#define XEC_EEPROM_DELAY_MS			5
+
+struct eeprom_xec_regs {
+	uint32_t MODE;
+	uint32_t EXECUTE;
+	uint32_t STATUS;
+	uint32_t INTR_ENABLE;
+	uint32_t PASSWORD;
+	uint32_t UNLOCK;
+	uint32_t LOCK;
+	uint32_t _reserved;
+	uint8_t BUFFER[32];
+};
+
+struct eeprom_xec_config {
+	struct eeprom_xec_regs * const regs;
+	size_t size;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
+};
+
+K_MUTEX_DEFINE(lock);
+
+static void eeprom_xec_execute_reg_set(struct eeprom_xec_regs * const regs,
+						uint32_t transfer_size, uint32_t command,
+						uint16_t eeprom_addr)
+{
+	uint32_t temp;
+
+	if (transfer_size == XEC_EEPROM_PAGE_SIZE) {
+		temp =  command +
+				(eeprom_addr & XEC_EEPROM_EXC_ADDR_MASK);
+	} else {
+		temp = (transfer_size << XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS) +
+				command + (eeprom_addr & XEC_EEPROM_EXC_ADDR_MASK);
+	}
```

Comment:
What do you think of below simplified logic? 

```
uint32_t temp = command + (eeprom_addr & XEC_EEPROM_EXC_ADDR_MASK);

if (transfer_size != XEC_EEPROM_PAGE_SIZE) {
	temp += (transfer_size << XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS);
}
```

Reply:
updated as per recommendation.